docs/specs/add-kubernetes-support.md

```diff
@@ -0,0 +1,247 @@
+Kubernetes is the upstream community for OpenShift and the broker should be
+consumuing the latest upstream code for development purposes.
+
+This change will improve the Broker development process by having the Broker
+constantly being developed and tested against the latest upstream code.
+
+The Broker's functionality won't change, but all the tooling around it will.
+
+## Split Out All Cluster logic
+OpenShift is Kubernetes plus a few features, API name changes, and tools.
+In the core ansible-service-broker code, ```pkg/cluster/kubernetes.go```
+should be able to account for about 80% of the cluster opertations required
+by the broker. The remaining ~20% will be held in ```pkg/cluster/openshift.go```.
+
+### Code
+Start by creating a new pkg for cluster objects, ```pkg/cluster/...```. From
+there, the cluster pkg will hold ```pkg/cluster/cluster.go```,
+```pkg/cluster/kubernetes.go```, and ```pkg/cluster/openshift.go```.
+
+The file ```pkg/cluster/cluster.go```  will provide an object that will abstract
+cluster logic.
+
+Using an abstration is an advantage because the broker code will only have a
+single path. It will call the cluster object asking for an operation. For
+example, to create a new Pod:
+```diff
+- k8scli.CoreV1().Pods(ns).Create(pod)
++ Cluster.CreatePod(name, image, extraVars, pullPolicy, serviceAccountName, ns)
+```
+
+The Cluster object will recieve the request and call the correct API.
+```diff
++ func CreatePod(name string, image spec.Image, extraVars string,
++                pullPolicy string, serviceAccountName string, ns string) {
++	pod := &v1.Pod{
++		ObjectMeta: metav1.ObjectMeta{
++			Name: apbID,
++		},
++		Spec: v1.PodSpec{
++			Containers: []v1.Container{
++				{
++					Name:  "apb",
++					Image: spec.Image,
++					Args: []string{
++						action,
++						"--extra-vars",
++						extraVars,
++					},
++					ImagePullPolicy: pullPolicy,
++				},
++			},
++			RestartPolicy:      v1.RestartPolicyNever,
++			ServiceAccountName: serviceAccountName,
++		},
++	}
++
++       k8scli.CoreV1().Pods(ns).Create(pod)
++
++ }
+
+```
+
+If there are identical API objects between COEs, check which cluster is
+being used.
+```diff
++ func CreateClusterRoleBinding(roleBindingName string, svcAccountName string,
++                               ApbRole string, ns string) {
++	roleBindingM := map[string]interface{}{
++		"apiVersion": "v1",
++		"kind":       "RoleBinding",
++		"metadata": map[string]string{
++			"name":      roleBindingName,
++			"namespace": ns,
++		},
++		"subjects": []map[string]string{
++			map[string]string{
++				"kind":      "ServiceAccount",
++				"name":      svcAccountName,
++				"namespace": ns,
++			},
++		},
++		"roleRef": map[string]string{
++			"name": ApbRole,
++		},
++      }
++
++      if OpenShift {
++         OCcli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
++      } elif Kubernetes {
++         k8scli.CoreV1().ClusterRoleBinding(ns).Create(roleBindingM)
```

Comment:
Yes, albeit a bit differently than traditional virtual methods. Go chooses to use composition over inheritance. On a high level, it's all about sharing behavior and delegation of specifics to components that know better (also providing encapsulation). We've organized our registry code and adapters in a similar way, so there is precedence, and I think we can leverage a similar pattern.

Spitballing, we can have a parent COE object that exposes the set of methods the broker needs to run APB actions. I would take the `RunPod` abstraction even further to `RunApbAction`. The fact that APBs are actually run as pods is an implementation detail; also, it provides optionality to run an APB action in a completely different way, (in an environment where the concept of pods don't even exist). Broker operates here on a higher level as well, calling into `coe.RunApbAction(ActionProvision, params)`.

Broker can build a coe runner via factory method, which is responsible for constructing a coe runner based on the current, concrete cluster the broker is configured to target: `runner := NewCOERunner(ClusterTypeKubernetes)`. Inside this, we can construct a parent COE with a `clusterDelegate ClusterDelegate` field, where ClusterDelegate is an interface sitting in front of cluster specifics. We can implement a `ClusterDelegate` for kubernetes and for openshift, and this is where the specialized behavior belongs.

Pseudocode:


```go
type ClusterDelegate interface {
    SetupPermissions()
}

func (r *COERunner) RunApbAction(...) {
    r.sharedBehavior(...)
    r.clusterDelegate.SetupPermissions()
}
```

In this case, we know we need to `SetupPermissions` for all cluster types, they just need to be done in a different way, which the delegates define themselves. For similar cases, think about what's actually getting achieved where differences arise, abstract and declare that behavior (`SetupPermissions`) on the `ClusterDelegate`, and then allow the specific implementations to define what it actually means to `SetupPermissions`.

More food for thought, this very much reminds me of PIMPL from the c++ world, or the Bridge Pattern:
http://en.cppreference.com/w/cpp/language/pimpl
https://en.wikipedia.org/wiki/Bridge_pattern